File: go-ms/domain/usecase/cases_use_case.go
package usecase

import (
	model "maocq/go-ms/domain/model/account"
	account "maocq/go-ms/domain/model/account/gateways"
	hello "maocq/go-ms/domain/model/hello/gateways"
	"time"
)

type CasesUseCase struct {
	AccountRepository account.AccountRepository
	HelloRepository   hello.HelloRepository
}

func (c *CasesUseCase) CaseOne(latency string) (*model.Account, error) {
	if _, err := c.HelloRepository.Hello(latency); err != nil {
		return nil, err
	}
	if _, err := c.HelloRepository.Hello(latency); err != nil {
		return nil, err
	}
	account := c.AccountRepository.FindById(4000)
	if _, err := c.HelloRepository.Hello(latency); err != nil {
		return nil, err
	}
	account.UpdateDate = time.Now()
	return c.AccountRepository.Update(account), nil
}

func (c *CasesUseCase) CaseTwo(latency string) ([]*model.Account, error) {
	accounts := c.AccountRepository.FindOne()
	if _, err := c.HelloRepository.Hello(latency); err != nil {
		return nil, err
	}
	return accounts, nil
}

func (c *CasesUseCase) CaseThree() *model.Account {
	return c.AccountRepository.FindById(4000)
}

File: java-reactive-ms/domain/usecase/src/main/java/com/maocq/usecase/cases/CasesUseCase.java
package com.maocq.usecase.cases;

import com.maocq.model.account.Account;
import com.maocq.model.account.gateways.AccountRepository;
import com.maocq.model.hello.gateways.HelloRepository;
import lombok.RequiredArgsConstructor;
import reactor.core.publisher.Flux;
import reactor.core.publisher.Mono;

import java.time.ZonedDateTime;

@RequiredArgsConstructor
public class CasesUseCase {

  private final AccountRepository accountRepository;
  private final HelloRepository helloRepository;

  public Mono<Account> caseOne(int latency) {
    return helloRepository.hello(latency)
        .flatMap(x -> helloRepository.hello(latency))
        .flatMap(x -> accountRepository.findById(4000)
        .flatMap(account -> helloRepository.hello(latency)
        .flatMap(y -> {
          account.setUpdateDate(ZonedDateTime.now());
          return accountRepository.update(account);
        })));
  }

  public Flux<Account> caseTwo(int latency) {
    return accountRepository.findOne()
        .flatMap(account -> helloRepository.hello(latency)
          .map(x -> account));
  }

  public Mono<Account> caseThree() {
    return accountRepository.findById(4000);
  }
}
